Add symfony mailer integration

bramtervoort opened the pull request:
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes #issuenum
| Related issues/PRs | #issuenum
| License | MIT

#### What's in this PR?

Created a mailer helper class to send emails using the new sf mailer
component. Created a configuration option to switch to varying mail
helpers.

#### Why?

It would be nice to be able to use the new Symfony Mailer component for sending the emails. 

#### Example Usage

In the configuration just state: 
```yml
sulu_form:
    mail:
        helper: "mailer"
```

Note that the user can now create customer helpers to and wire them in using a tag and this configuration setting.

#### To Do

- [ ] Create documentation for the new tag i added if that is needed in you opinion.

alexander-schranz commented on Resources/config/services.xml:
```diff
@@ -15,20 +15,38 @@
             <argument>%sulu_form.mail.to%</argument>
             <argument>%sulu_form.mail.sender%</argument>
             <argument type="service" id="logger" />
+            <tag name="sulu_mail.handler" key="SwiftMailer" />
```
I would not go here with tag I would implement it the same way we did implement the storage adapter in sulu.

alexander-schranz commented on Resources/config/services.xml:
```diff
@@ -15,20 +15,38 @@
+        </service>
+
         <!-- Mail-->
         <service id="sulu.mail.helper" class="Sulu\Bundle\FormBundle\Mail\Helper">
```
I would change here the service id to `sulu.mail.swift_mailer`

We will create an alias based on the configuration a service alias `sulu.mail.helper`for the configured service which by default will be `swift_mailer`

alexander-schranz commented on Resources/config/services.xml:
```diff
@@ -15,20 +15,38 @@
+            <tag name="sulu_mail.handler" key="SwiftMailer" />
+        </service>
+
+        <service id="sulu.mail.mailer_helper" class="Sulu\Bundle\FormBundle\Mail\MailerHelper">
```
This should be then `sulu.mail.helper.swift_mailer.mailer`

alexander-schranz commented on Resources/config/services.xml:
```diff
@@ -15,20 +15,38 @@
+            <argument>%sulu_form.mail.to%</argument>
+            <argument>%sulu_form.mail.sender%</argument>
+            <argument type="service" id="logger" />
+            <tag name="sulu_mail.handler" key="MailerHelper" />
```
Tag can be removed

alexander-schranz commented on Resources/config/services.xml:
```diff
@@ -15,20 +15,38 @@
 
         <service id="sulu_mail.null_helper" class="Sulu\Bundle\FormBundle\Mail\NullHelper">
             <argument type="service" id="logger" />
+            <tag name="sulu_mail.handler" key="NullHelper" />
```
Tag can be removed

alexander-schranz commented on Resources/config/services.xml:
```diff
@@ -15,20 +15,38 @@
+            <tag name="sulu_mail.handler" key="MailerHelper" />
         </service>
 
         <service id="sulu_mail.null_helper" class="Sulu\Bundle\FormBundle\Mail\NullHelper">
```
Don't see any usage for it, seems like just a service which was not used. I would `deprecate` the service definition and the whole NullLogger class itself.

alexander-schranz commented on Resources/doc/index.md:
````diff
@@ -46,6 +46,16 @@ sulu_form:
+```yml
+sulu_form:
+    mail:
+        helper: "SwiftMailer"
````
Available options will then be `swift_mailer` or `mailer`

alexander-schranz replied:
```suggestion
        helper: "swift_mailer" # is default
```

alexander-schranz commented on Tests/Application/.env:
```diff
@@ -2,3 +2,4 @@ APP_ENV=test
 DATABASE_URL=mysql://root:@127.0.0.1:3306/su_form_test
 DATABASE_CHARSET=utf8mb4
 DATABASE_COLLATE=utf8mb4_unicode_ci
+MAIL_HANDLER=SwiftMailer
```
would just go here with:

```
MAIL_HANDLER=
```

if not set it should use the default in your SuluForm Configuration class

alexander-schranz commented on Mail/MailerHelper.php:
```diff
@@ -0,0 +1,258 @@
+     * @param Address[] $bccMail list of addresses already accounting for the defaults may be an empty array
+     * @param Address|null $sender address already accounting for the defaults
+     */
+    protected function setHeaders(
```
make `private` not sure if this will change so I can not give here any BC promise

alexander-schranz replied:
still valid this should be private as we can currently not guarente a BC here.

alexander-schranz commented on Mail/MailerHelper.php:
```diff
@@ -0,0 +1,258 @@
+     * @param Email $message
+     * @param \SplFileInfo[] $attachments
+     */
+    protected function setAttachments(
```
same as above should be private

alexander-schranz replied:
same as above still valid the protected function should be private

alexander-schranz commented on Mail/MailerHelper.php:
```diff
@@ -0,0 +1,258 @@
+     * @param string $body
+     * @param string $plainText
+     */
+    protected function setBody(
```
same as above should be private

alexander-schranz replied:
same as above still valid the protected function should be private we can not give bc support for them here.

alexander-schranz commented on Mail/HelperAwareInterface.php:
```diff
@@ -0,0 +1,10 @@
+namespace Sulu\Bundle\FormBundle\Mail;
+
+
+interface HelperAwareInterface
```
I would not go with a AwareInterface and setter so this can be removed

alexander-schranz commented on Mail/HelperAwareConfigurator.php:
```diff
@@ -0,0 +1,33 @@
+namespace Sulu\Bundle\FormBundle\Mail;
+
+
+class HelperAwareConfigurator
```
I would not go with a Configurator here so this can be removed.

alexander-schranz commented on Form/Handler.php:
```diff
@@ -83,6 +86,11 @@ public function __construct(
         $this->honeyPotField = $honeyPotField;
     }
 
+    public function setHelper(HelperInterface $mailHelper): void
```
As written beneath we will inject the correct `Mail\HelperInterface $mailHelper` in the constructor so no setter is needed here.

alexander-schranz commented on Form/Handler.php:
```diff
@@ -28,7 +30,7 @@
  * Handling of form based on form configuration.
  */
-class Handler implements HandlerInterface
+class Handler implements HandlerInterface, HelperAwareInterface
```
also this Interface can be removed

alexander-schranz commented on DependencyInjection/SuluFormExtension.php:
```diff
@@ -207,5 +209,7 @@ public function load(array $configs, ContainerBuilder $container)
                 ->setPublic(true);
         }
+
+        $container->setParameter('sulu_mail.mail.helper_name', $config['mail']['helper']);
```
We need to check if the `mailer` component is installed so we should go here with something like:

```php
$defaultMailerHelper = null;
if (class_exists(SwiftMailer::class)) {
      $defaultMailerHelper = 'swift_mailer';
      $loader->load('swift_mailer.xml'); // this file contains the swift mailer configuration
}

if (class_exists(Mailer::class)) {
      $defaultMailerHelper = $defaultMailer ?: 'mailer';
      $loader->load('mailer.xml'); // this file contains the mailer configuration
}
```

Instead of setting a parameter here we will create a service alias:

```php
$mailHelper = $container->resolvePlaceholder($config['mail']['helper'], true);
if ('auto' === 'mailerHelper') {
      $mailerHelper = $defaultMailerHelper;
}

container->setAlias('sulu.mail.helper', 'sulu.mail.helper.' . $mailHelper)
                ->setPublic(true);

container->setAlias('sulu.mail.helper', 'sulu.mail.helper.' . $mailHelper)
                ->setPublic(true);
```

alexander-schranz commented on Mail/MailerHelper.php:
```diff
@@ -0,0 +1,258 @@
+use Symfony\Component\Mime\Address;
+use Symfony\Component\Mime\Email;
+
+class MailerHelper implements HelperInterface
```
Will have a deeper look at this later.

alexander-schranz commented on Resources/config/services.xml:
```diff
@@ -15,20 +15,38 @@
             <argument type="service" id="sulu_media.media_manager" />
             <argument>%sulu_form.honeypot_strategy%</argument>
             <argument>%sulu_form.honeypot_field%</argument>
+            <configurator service="sulu.mail.helper_aware_configurator" />
```
As written beneath this can be removed with the suggestions

alexander-schranz commented on Resources/config/services.xml:
```diff
@@ -15,20 +15,38 @@
         </service>
         <service id="Sulu\Bundle\FormBundle\Form\HandlerInterface" alias="sulu_form.handler"/>
 
+        <service id="sulu.mail.helper_aware_configurator" class="Sulu\Bundle\FormBundle\Mail\HelperAwareConfigurator">
```
As written beneath this can be removed with the suggestions

alexander-schranz replied:
Should still be valid this should be able to be removed with setting the alias like we do it in the other cases.

alexander-schranz commented on Tests/Functional/Listener/FormRequestTest.php:
```diff
@@ -0,0 +1,238 @@
+use Sulu\Bundle\FormBundle\Entity\FormTranslationReceiver;
+use Sulu\Bundle\TestBundle\Testing\SuluTestCase;
+
+class FormRequestTest extends SuluTestCase
```
Nice 👍

alexander-schranz commented on Tests/Functional/Listener/FormRequestTest.php:
```diff
@@ -0,0 +1,238 @@
+
+    public function testUsesMailerWhenMailerHandlerSet()
+    {
+        $_ENV['MAIL_HANDLER'] = 'MailerHelper';
```
would not set a `$_ENV` here would create seperate TestCase for the `Mailer` and set there the KernelConfiguration `environment`: https://github.com/sulu/sulu/blob/3b41b627fbc1cf93be7f98eb3e3d4ebe35883b35/src/Sulu/Bundle/TestBundle/Testing/KernelTrait.php#L66-L69

to something like `test_mailer` and then in then create a `test_mailer/config.yaml` or `config_test_mailer.yaml` which is loaded by the test kernel where:

```
sulu_form:
     mail:
           helper: 'mailer'
```

is configured

alexander-schranz commented on Tests/Functional/Listener/FormRequestTest.php:
```diff
@@ -0,0 +1,238 @@
+    }
+
+
+    private function createFullForm(): Form
```
Would move this into a `CreateFormTrait` in Tests/Functional/Trait namespace.

bramtervoort replied:
I turned this into a fixture instead since I think that is more close to what it should be. Of course happy to still make it into a trait.

I remember there was a nice way to apply fixtures but well i came up with this :)

alexander-schranz commented on Tests/Functional/Listener/FormRequestTest.php:
```diff
@@ -0,0 +1,238 @@
+        $this->assertResponseRedirects('?send=true');
+    }
+
+    private function inititalize(): void
```
with splitting this into multiple TestCase this can now be the `setUp`

alexander-schranz commented on Tests/Unit/Mail/MailerHelperTest.php:
```diff
@@ -0,0 +1,218 @@
+
+    protected function setUp(): void
+    {
+        $this->mailerMock = $this->getMockBuilder(MailerInterface::class)->getMock();
```
We use `prophecy` to for mocking so I would keep that. Ping me or contact me on slack if you need any help.

alexander-schranz commented on composer.json:
```diff
@@ -17,18 +17,20 @@
         "symfony/event-dispatcher-contracts": "^1.0 || ^2.0",
         "symfony/form": "^4.3 || ^5.0",
-        "symfony/validator": "^4.3 || ^5.0",
+        "symfony/mime": "^4.4 || ^5.1",
+        "symfony/mailer": "^4.4 || ^5.1",
```
We should not increase the dependency tree and as sulu ships currently with swiftmailer. So currently this should be just `dev` dependencies and if somebody want to use the `Mailer` implementation he need to install the symfony mailer component.

alexander-schranz replied:
This should still be "dev" dependencies at current state mailer should not be shipped with sulu form bundle. When somebody want to switch to the mailer it should actively require it in there composer.json file.

Also it should be here `^4.4 || ^5.0`

alexander-schranz reviewed (changes requested):
@bramtervoort Thank you for working on this 🚀 . I think we should use the same pattern which we currently use in sulu for this kind of things, else it is getting confused having different implementation for the same kind of thing. I added some comments and did not yet have enough time to have a deeper look at the tests and the implementation but wanted to give you feedback as soon as it was possible.

bramtervoort commented:
Thank you Alexander for your feedback. It will be a couple of day's before I have the time to resolve them properly.

alexander-schranz commented:
@bramtervoort no problem. Feel free to ping me here or over slack if you need any help.

bramtervoort commented:
i made some progress with this. next weekend ill be finishing this.

bramtervoort commented:
There all refactored, i just love that `prophecy`. I agree it looks much better now. 

I refactored the form to a fixture if you don't agree I will be happy to turn it into a trait of course. Took some time to finish this all but i hope you can still have a look at it. Happy to resolve more review issues.

alexander-schranz commented on DependencyInjection/Configuration.php:
```diff
@@ -22,6 +23,10 @@
  */
 class Configuration implements ConfigurationInterface
 {
+    const NULL_HELPER = 'NullHelper';
```
should be removed as null://localhost should be configured then as MAILER DSN and adding here the null helper into the new configuration doesn't make any sense at this state

alexander-schranz commented on DependencyInjection/Configuration.php:
```diff
@@ -22,6 +23,10 @@
 class Configuration implements ConfigurationInterface
 {
+    const NULL_HELPER = 'NullHelper';
+    const SWIFT_MAILER_HELPER = 'SwiftMailer';
```
```suggestion
    const SWIFT_MAILER_HELPER = 'swift_mailer';
```

alexander-schranz commented on DependencyInjection/Configuration.php:
```diff
@@ -22,6 +23,10 @@
 {
+    const NULL_HELPER = 'NullHelper';
+    const SWIFT_MAILER_HELPER = 'SwiftMailer';
+    const MAILER_HELPER = 'Mailer';
```
```suggestion
    const MAILER_HELPER = 'mailer';
```

alexander-schranz commented on DependencyInjection/SuluFormExtension.php:
```diff
@@ -207,5 +211,25 @@ public function load(array $configs, ContainerBuilder $container)
+            $container->setAlias('sulu.mail.helper', 'sulu_mail.null_helper');
+        } elseif ($helper == Configuration::SWIFT_MAILER_HELPER) {
+            $container->setAlias('sulu.mail.helper', 'sulu.mail.swift_mailer');
+        }
```
This can be simplified to:

```suggestion
        $container->setAlias('sulu.mail.helper', 'sulu.mail.' . $helper);
```

This make it also extendable for the third party provide there own mailer.

alexander-schranz commented on DependencyInjection/SuluFormExtension.php:
```diff
@@ -207,5 +211,25 @@ public function load(array $configs, ContainerBuilder $container)
+        if (\method_exists($container, 'resolveEnvPlaceholders')) {
+            $helper = $container->resolveEnvPlaceholders($helper, true);
+        }
+        $container->setParameter('sulu.mail.helper', $helper);
```
This parameter should be removed. It is not used by us and the dev should not have to care about which mailer is used to send the emails.

alexander-schranz commented on Form/Builder.php:
```diff
@@ -272,4 +274,5 @@ private function getWebspaceKey(): ?string
 
         return $webspaceKey;
     }
+
```
should be removed

alexander-schranz commented on Event/RequestListener.php:
```diff
@@ -64,7 +65,6 @@ public function onKernelRequest(RequestEvent $event): void
         }
 
         $request = $event->getRequest();
-
```
would not remove this empty line :)

alexander-schranz commented on DependencyInjection/Configuration.php:
```diff
@@ -51,6 +56,9 @@ public function getConfigTreeBuilder()
                 ->children()
+                    ->scalarNode('helper')
+                        ->defaultValue(self::SWIFT_MAILER_HELPER)
+                        ->end()
```
end should be on the same level as the opening node.

```suggestion
                    ->end()
```

alexander-schranz commented on Mail/MailerHelper.php:
```diff
@@ -0,0 +1,258 @@
+        );
+
+        $this->mailer->send($message);
+        return 0;
```
Empty line should be here before the return.

```suggestion

        return 0;
```

alexander-schranz commented on Resources/config/services.xml:
```diff
@@ -18,17 +18,34 @@
+
         <!-- Mail-->
-        <service id="sulu.mail.helper" class="Sulu\Bundle\FormBundle\Mail\Helper">
+        <service id="sulu.mail.helper" alias="sulu.mail.swift_mailer" />
```
This is set in the configuration so this don't need to be set here again. So we can remove this.

alexander-schranz commented on Resources/config/services.xml:
```diff
@@ -18,17 +18,34 @@
+
         <service id="sulu_mail.null_helper" class="Sulu\Bundle\FormBundle\Mail\NullHelper">
             <argument type="service" id="logger" />
+            <tag name="sulu_mail.handler" key="NullHelper" />
```
Tag can be removed but service should also be marked as deprecated as it was actually never used:

```suggestion
           <deprecated />
```

alexander-schranz commented on Resources/doc/index.md:
````diff
@@ -46,6 +46,16 @@ sulu_form:
         sender: "%env(SULU_ADMIN_EMAIL)%"
 ```
 
+Configure the email handler to SwiftMailer (default), 
````
Configure the email handler to SwiftMailer (`swift_mailer`) or the Symfony Mailer (`mailer`)

alexander-schranz commented on SuluFormBundle.php:
```diff
@@ -13,8 +13,10 @@
 
 use Sulu\Bundle\FormBundle\DependencyInjection\CompilerPass\DynamicListBuilderCompilerPass;
 use Sulu\Bundle\FormBundle\DependencyInjection\CompilerPass\ListProviderCompilerPass;
+use Sulu\Bundle\FormBundle\DependencyInjection\CompilerPass\MailHelperCompilerPass;
```
looks like unused use statements here

alexander-schranz commented on Tests/Application/Kernel.php:
```diff
@@ -11,9 +11,13 @@
 
 namespace Sulu\Bundle\FormBundle\Tests\Application;
 
+use Sulu\Bundle\FormBundle\Tests\Application\Controller\DefaultController;
```
looks like unused use statements here

alexander-schranz commented on Tests/Application/MailerKernel.php:
```diff
@@ -0,0 +1,26 @@
+<?php
+
+
+namespace Sulu\Bundle\FormBundle\Tests\Application;
```
code style should be fixed and the copyright header should be added

alexander-schranz commented on Tests/Application/MailerKernel.php:
```diff
@@ -0,0 +1,26 @@
+            . 'var' . \DIRECTORY_SEPARATOR
+            . 'cache' . \DIRECTORY_SEPARATOR
+            . $this->getContext()  . \DIRECTORY_SEPARATOR
+            . $this->environment;
```
this should have its own cache directory e.g. `$this->environment . '_mailer'` else it could conflict with the other Kernel.

alexander-schranz commented on Tests/Application/config/routing_website.yml:
```diff
@@ -1,2 +0,0 @@
-sulu_form:
```
why was this removed?

alexander-schranz commented on Tests/Application/config/config_test_mailer.yaml:
```diff
@@ -0,0 +1,6 @@
+#
+sulu_form:
+    mail:
+        helper: 'Mailer'
```
```suggestion
sulu_form:
    mail:
        helper: 'mailer'
```

alexander-schranz commented on Tests/Functional/Mail/HelperTestCase.php:
```diff
@@ -0,0 +1,107 @@
+        $fixture->load($this->entityManager);
+    }
+
+//    public function testPageRendersWithoutForm()
```
should be removed when not used

alexander-schranz commented on Tests/Functional/Mail/Fixtures/LoadFormFixture.php:
```diff
@@ -0,0 +1,117 @@
+        $formField->setForm($form);
+
+        $manager->persist($form);
+        $manager->flush();
```
We should not flush here. the flush and clear should do the test case itself as maybe there other entities need to be flushed with it and as flush has a big impact on performance we should not use only use it when it is really needed.

alexander-schranz commented on Tests/Functional/Mail/Fixtures/LoadFormFixture.php:
```diff
@@ -0,0 +1,117 @@
+        $manager->flush();
+        $manager->clear();
+    }
+
```
code style should also be fixed here.

alexander-schranz commented on Tests/Functional/Mail/HelperTest.php:
```diff
@@ -0,0 +1,32 @@
+<?php
+
```
PHPCS should be run here not sure why php cs was not run for this PR. this files should be fixed when run:

```
composer php-cs-fix
```

alexander-schranz commented on Tests/Unit/Mail/MailerHelperTest.php:
```diff
@@ -0,0 +1,210 @@
+ * Class MailerHelperTest
+ * This tests the MailerHelper class that sends an email using the symfony mailer component.
+ *
+ * The MailerHelper class has 3 protected methods that could be overwritten to change the mail message.
```
I will currently not support this. If somebody want to hook in he should use the official symfony way, listening on events of symfony mailer or create a swift mailer plugin.

alexander-schranz reviewed (commented):
@bramtervoort Thx for working on this. I added some more comments about the actual implementation.

alexander-schranz commented:
@bramtervoort can you also rebase this branch to solve the conflicts and can you try to update your forks branches. I'm not sure why github action are not started for this PR but maybe it has todo that the fork branches are not up2date think it is worse a try.

bramtervoort commented:
> @bramtervoort Thx for working on this. I added some more comments about the actual implementation.

Thx, I will check if can find a little time to work some more on this.

Commit:
added helper support sf mailer and configuration mechanism to switch

Created a mailer helper class to send emails using the new sf mailer
component. Created a configuration option to switch to varying mail
helpers.

Commit:
refactor configuration of helpers

refactored them as commented in the reviews

Commit:
fix code style and review

Commit:
feature/sf-mailer-in-form-bundle added new form template

Commit:
fixed the unit tests and code style

bramtervoort commented:
There found the time to update this :D thank you for your patience. If i can do anything to fix or improve the code please let me know. ( sorry this took me so long ;) )

alexander-schranz committed:
Some minor changes for the mail helpers

alexander-schranz commented:
@bramtervoort Thank you for contributing this 👍
